podman rm/stop --cidfile

vrothberg opened the pull request:
Add a --cidfile flag to podman rm/stop to pass a container ID via a
file.  Podman run already provides the functionaly to store the ID
in a specified file which we now complete with rm/stop.  This allows
for a better life-cycle management in systemd services.

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

vrothberg commented:
@rhatdan @giuseppe @edsantiago PTAL

mheon commented:
What's the intent here? I don't mind the change, but it seems a little strange at first glance

vrothberg commented:
> What's the intent here? I don't mind the change, but it seems a little strange at first glance

That's needed for the _new_ systemd services. `ExecStart` writes the file, `ExecStop` reads it. This way, we don't need to hard-code any IDs or names in the files and make them shareable. I can demo it on Monday.

rhatdan commented on cmd/podman/common.go:
```diff
@@ -39,24 +39,45 @@ func shortID(id string) string {
+	specifiedAll, _ := c.Flags().GetBool("all")
+	specifiedLatest, _ := c.Flags().GetBool("latest")
+	specifiedCIDFile := false
+	if cid, _ := c.Flags().GetString("cidfile"); cid != "" {
```
Why can't you just check for this in the same line that you lookup all and latest?

vrothberg replied:
`all` and `latest` are bools, while `cidfile` is a string. So we first need to lookup the string and then check if it's empty or not.

TomSweeneyRedHat commented on cmd/podman/common.go:
```diff
@@ -39,24 +39,45 @@ func shortID(id string) string {
-func checkAllAndLatest(c *cobra.Command, args []string, ignoreArgLen bool) error {
+// checkAllAndLatest checks that --all and --latest are used correctly. If
+// cidfile is set, also check for the --cidfile flag.
+func checkAllAndLatest(c *cobra.Command, args []string, ignoreArgLen bool, cidfile bool) error {
```
might be good to massage Cid into the function name.  perhaps checkCidAllAndLatest

baude replied:
what about a specific function for it that wraps checkAllAndLatest?  That allows you to address Tom's comment and not change other usages? Only a suggestion.  Defer to your judgement,

baude replied:
thinking about this more ... it might be a good idea because it would allow you to check for colliding flags, etc outside the checkAllAndLatest ... again, your call.

vrothberg replied:
I guess I'll go the lazy approach and follow `--storage` which also conflicts with `--all` and `--latest` but does the check outside the function. It's frustrating the `cobra` does not allow to specify conflicting flags. Maybe worth contributing?

vrothberg replied:
Eek, that doesn't work nicely either. Same applies to a wrapper func as we really need to check all flags together at once. It renamed the func though. Note that the additional parameter is needed for commands that don't support --cidfile ... it avoids printing errors that include --cidfile.

TomSweeneyRedHat commented:
and I neglected to add you need some bash completions too.

baude commented:
a couple of minor suggestions @vrothberg ... great job and i like it

giuseppe commented:
code LGTM.

Just a question, differently than `create` and `run`, `stop` and `rm` accept multiple IDs.

Could we change `CIDFile` to be a string slice, so that it is possible to do something like?

`podman stop --cidfile /tmp/cid --cidfile /tmp/cid2 ...`

vrothberg commented:
> Could we change `CIDFile` to be a string slice, so that it is possible to do something like?
> 
> `podman stop --cidfile /tmp/cid --cidfile /tmp/cid2 ...`

Sure, I can extend the PR to cover this case as well, both for `rm` and `stop`.

vrothberg commented:
Updated:
* `--cidfile` can now be specified multiple times. Tests have been updated accordingly.
* Added bash-completions.
* Updated man pages.
* s/checkAllAndLatest/checkAllLatestAndCIDFile/

vrothberg commented:
PTanotherL @TomSweeneyRedHat @baude @giuseppe

TomSweeneyRedHat commented:
LGTM assuming happy tests.

giuseppe commented:
/lgtm
